internal/fs/vss_windows.go

```diff
@@ -0,0 +1,1051 @@
+		return VssSnapshot{}, err
+	}
+
+	if hRes, isSupported := iVssBackupComponents.IsVolumeSupported(volume); hRes != S_OK {
+		return VssSnapshot{}, fmt.Errorf(
+			"VSS error: IsVolumeSupported() returned %s (%#x)",
+			hRes.Str(), hRes,
+		)
+	} else if !isSupported {
+		return VssSnapshot{}, fmt.Errorf(
+			"VSS error: Snapshots are not supported for volume %s",
+			volume,
+		)
+	}
+
+	// TODO:
+	//   What about IVssBackupComponents::AddComponent?
+	//   see https://docs.microsoft.com/en-us/windows/desktop/VSS/simple-shadow-copy-creation-for-backup
+
+	hRes, snapshotSetID := iVssBackupComponents.StartSnapshotSet()
+	if hRes != S_OK {
+		return VssSnapshot{}, fmt.Errorf(
+			"VSS error: StartSnapshotSet() returned %s (%#x)", hRes.Str(), hRes,
+		)
+	}
+
+	if hRes := iVssBackupComponents.AddToSnapshotSet(volume, &snapshotSetID); hRes != S_OK {
+		return VssSnapshot{}, fmt.Errorf(
+			"VSS error: AddToSnapshotSet() returned %s (%#x)", hRes.Str(), hRes,
+		)
+	}
+
+	mountPoints, err := enumerateMountedFolders(volume)
+	if err != nil {
+		return VssSnapshot{}, fmt.Errorf(
+			"VSS error: failed to enumerate mount points for volume %s: %s",
+			volume, err,
+		)
+	}
+
+	mountPointInfo := make(map[string]MountPoint)
+
+	for _, mountPoint := range mountPoints {
+		// ensure every mountpoint is available even without a valid
+		// snapshot because we need to consider this when backing up files
+		mountPointInfo[mountPoint] = MountPoint{
+			isSnapshotted: false,
+		}
+
+		if hRes, isSupported := iVssBackupComponents.IsVolumeSupported(mountPoint); hRes != S_OK {
+			continue
+		} else if !isSupported {
+			continue
+		}
+
+		var mountPointSnapshotSetID ole.GUID
+		if hRes := iVssBackupComponents.AddToSnapshotSet(mountPoint, &mountPointSnapshotSetID); hRes != S_OK {
+			return VssSnapshot{}, fmt.Errorf(
+				"VSS error: AddToSnapshotSet() returned %s (%#x) when adding mount point %s",
+				hRes.Str(), hRes, mountPoint,
+			)
+		}
+
+		mountPointInfo[mountPoint] = MountPoint{
+			isSnapshotted: true,
+			snapshotSetID: mountPointSnapshotSetID,
+		}
+	}
+
+	err = handleAsyncFunctionCall(
+		iVssBackupComponents.PrepareForBackup,
+		"PrepareForBackup", timeoutInMillis,
+	)
+	if err != nil {
+		return VssSnapshot{}, err
+	}
+
+	err = handleAsyncFunctionCall(
+		iVssBackupComponents.DoSnapshotSet, "DoSnapshotSet", timeoutInMillis,
+	)
+	if err != nil {
+		return VssSnapshot{}, err
+	}
+
+	var snapshotProperties VssSnapshotProperties
+	if hRes := iVssBackupComponents.GetSnapshotProperties(
+		snapshotSetID, &snapshotProperties,
+	); hRes != S_OK {
+		return VssSnapshot{}, fmt.Errorf(
+			"VSS error: GetSnapshotProperties() returned %s (%#x)",
+			hRes.Str(), hRes,
+		)
+	}
+
+	for mountPoint, info := range mountPointInfo {
+		if hRes := iVssBackupComponents.GetSnapshotProperties(
+			info.snapshotSetID, &info.snapshotProperties,
+		); hRes != S_OK {
+			return VssSnapshot{}, fmt.Errorf(
+				"VSS error: GetSnapshotProperties() returned %s (%#x)",
```

Comment:
You right, it not fail at AddToSnapshotSet, but it fail at GetSnapshotProperties.
Simple example:
mountvol c:\tmp \\?\Volume{_this is volume D:_}\
chkdsk /F /V /X D: _(flag /X lock/unmount/whatever volume D:)_
restic backup --use-windows-vss c:\test _(backup unrelated file)_
repository d751adc4 opened successfully, password is correct
creating VSS snapshot for [c:\]
failed to create snapshot for [c:\]: VSS error: GetSnapshotProperties() returned E_INVALIDARG (0x80070057)
error: open c:\test\test.txt: The process cannot access the file because it is being used by another process.

Same setup, but old version of this PR without mountpoints:
repository d751adc4 opened successfully, password is correct
creating VSS snapshot for [c:\]
Files:           1 new,     0 changed,     0 unmodified
Dirs:            0 new,     1 changed,     0 unmodified
Added to the repo: 0 B
processed 1 files, 4 B in 0:03
snapshot e912af30 saved

Reply:
Resolved in https://github.com/restic/restic/pull/2274/commits/ff23626154403694361b53e29bf58e5e96e75100